spyder_kernels/customize/spydercustomize.py

```diff
@@ -867,8 +867,62 @@ def get_debugger(filename):
     return debugger, filename
 
 
+class FileNamespace():
+    """
+    Get a namespace and handle __file__ correctly
+    """
+
+    def __init__(self, filename, namespace=None, current_namespace=False):
+        self.filename = filename
+        self.namespace = namespace
+        self.current_namespace = current_namespace
+        self._previous_filename = None
+        self._previous_main = None
+        self._reset_main = False
+
+    def __enter__(self):
+        """
+        Get namespace and create __file__
+        """
+        # Save previous __file__
+        if self.namespace is None:
+            if self.current_namespace:
+                self.namespace = _get_globals()
+            else:
+                ipython_shell = get_ipython()
+                main_mod = ipython_shell.new_main_mod(
+                    self.filename, '__main__')
+                self.namespace = main_mod.__dict__
+                # Needed to allow pickle to reference main
+                if '__main__' in sys.modules:
+                    self._previous_main = sys.modules['__main__']
+                sys.modules['__main__'] = main_mod
+                self._reset_main = True
+        if '__file__' in self.namespace:
+            self._previous_filename = self.namespace['__file__']
+        self.namespace['__file__'] = self.filename
+        return self.namespace
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        Reset namespace['__file__']
+        """
+        if not self.current_namespace:
+            get_ipython().user_ns.update(self.namespace)
+        if self._previous_filename:
+            self.namespace['__file__'] = self._previous_filename
+        elif '__file__' in self.namespace:
+            # Avoid error when running `%reset -f` programmatically
+            # See issue spyder-ide/spyder-kernels#91
+            self.namespace.pop('__file__')
```

Comment:
Please enclose this in a try/except KeyError (as before) to avoid the issue referenced above this line.

Reply:
Isn't `if '__file__' in self.namespace` enough? I thought the `KeyError` was to catch that.

Reply:
`%reset -f` removes it after `runfile` is executed, so this could still happen (I haven't tested it with this PR though). Also, since you left the comment above, I thought it still happened.

Reply:
I left the comment to explain the `if '__file__' in self.namespace`. I tried spyder-ide/spyder-kernels#91 and it works as expected